DROOLS-5131: [DMN Designer] Boxed List support

manstis opened the pull request:
See https://issues.redhat.com/browse/DROOLS-5131

New grid type for "Lists".... 

I had to move both `Relation` and the new `List` to contain a `List<HasExpression>` rather than `List<Expression>` to be able to maximise re-use of existing code to get and set an expression _on_ an object instance so this PR is bloated with related changes for `Relation`.

jomarko commented:
Review blocked by https://issues.redhat.com/browse/DROOLS-5170

jomarko commented:
2 sonar cloud bugs of level C, one seems false positive, but one seems worth to have a look on
https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&open=AXDkahAUB2F7FzBhpmWB&pullRequest=3220&resolved=false&types=BUG

manstis commented:
@jomarko Stack-overflow issue fixed. @karreiro I've added the same header details as requested (good spot!). The header does not exist when the `List` is nested (for example in a `ContextEntry`).

manstis commented:
@jomarko I think _both_ SonarCloud "Level C" bugs are false positives:

1) [ecv.getValue().get()](https://sonarcloud.io/code?id=org.kie.workbench%3Akie-wb-common&pullRequest=3220&selected=org.kie.workbench%3Akie-wb-common%3Akie-wb-common-dmn%2Fkie-wb-common-dmn-client%2Fsrc%2Fmain%2Fjava%2Forg%2Fkie%2Fworkbench%2Fcommon%2Fdmn%2Fclient%2Feditors%2Fexpressions%2Ftypes%2Flist%2FListGrid.java&line=205)
2) [getExpression().get().get()](https://sonarcloud.io/code?id=org.kie.workbench%3Akie-wb-common&pullRequest=3220&selected=org.kie.workbench%3Akie-wb-common%3Akie-wb-common-dmn%2Fkie-wb-common-dmn-client%2Fsrc%2Fmain%2Fjava%2Forg%2Fkie%2Fworkbench%2Fcommon%2Fdmn%2Fclient%2Feditors%2Fexpressions%2Ftypes%2Flist%2FListGrid.java&line=288)

There have been some significant changes (in a new commit) since the last build. Let's see what SonarCloud makes of the revised PR and address as applicable thereafter.

jomarko commented:
jenkins execute full downstream build

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/list/MoveListRowsCommandTest.java:
```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
2020

manstis replied:
Done.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/list/ListExpressionColumnHeaderMetaDataTest.java:
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
```
2020

manstis replied:
Done.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/list/ListGrid.java:
```diff
@@ -0,0 +1,337 @@
+    @Override
+    public ListUIModelMapper makeUiModelMapper() {
+        return new ListUIModelMapper(this,
+                                     this::getModel,/**/
```
Done.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/list/ListGrid.java:
```diff
@@ -0,0 +1,337 @@
+
+    @Override
+    public void onItemSelected(final ListSelectorItem item) {
+        final ListSelectorTextItem li = (ListSelectorTextItem) item;
```
Yes, as the clickable items added to it are all `ListSelectorTextItem`

See [these](https://github.com/kiegroup/kie-wb-common/pull/3220/files/19205611aeb71a020ddcc44bbecc4dcc7a7ee2c7#diff-dcbffad8affcbfecf7bc4b8a18a8a68fR196-R208) lines.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/list/ListEditorDefinition.java:
```diff
@@ -0,0 +1,152 @@
+
+    @Override
+    public Optional<List> getModelClass() {
+        return Optional.of(new List());
```
Good question... it is used as the expression instance _behind_ the grid and hence needs to be a new instance each time... BUT the `class` is also used to lookup an editor for the expression instance. See `ExpressionEditorDefinitions`. I guess these two concerns should be separated and `ExpressionEditorDefinition` have two methods; one to return a `Class` used to lookup editors; and another to provide a new instance of an expression for editors.

WDYT? Adding it to https://issues.redhat.com/browse/DROOLS-2974 would be best.

If you agree I can do the necessary admin.

manstis commented:
@jomarko I have addressed your comments in a new commit. 

I have also added an integration test for client-side marshalling of `List`.

@jomarko @karreiro This PR can be re-reviewed at your leisure. Thank-you.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/src/test/resources/org/kie/workbench/common/dmn/showcase/client/selenium/DROOLS-5131 (List expression).xml:
```diff
@@ -0,0 +1,52 @@
+                 xmlns:dmndi='http://www.omg.org/spec/DMN/20180521/DMNDI/'
+                 xmlns:dc='http://www.omg.org/spec/DMN/20180521/DC/'
+                 id='_15712B07-1327-478A-A1A8-214878C1F307'
+                 name='DROOLS-5131'
```
I did do when I first wrote the _suite_ of tests; but subsequent ones have the JIRA number in their name.. happy to do either way. I don't mind; feel free to state your preference.

jomarko replied:
ok we mix already naming, so feel free to keep as it is
